x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -318,11 +320,40 @@ export class SessionIndex {
     const sessionIndexTemplateName = `${this.options.kibanaIndexName}_security_session_index_template_${SESSION_INDEX_TEMPLATE_VERSION}`;
     return (this.indexInitialization = new Promise<void>(async (resolve, reject) => {
       try {
+        // Check if legacy index template exists, and remove it if it does.
+        let legacyIndexTemplateExists = false;
+        try {
+          legacyIndexTemplateExists = (
+            await this.options.elasticsearchClient.indices.existsTemplate({
+              name: sessionIndexTemplateName,
+            })
+          ).body;
+        } catch (err) {
+          this.options.logger.error(
+            `Failed to check if session legacy index template exists: ${err.message}`
+          );
+          return reject(err);
+        }
+
+        if (legacyIndexTemplateExists) {
+          try {
+            await this.options.elasticsearchClient.indices.deleteTemplate({
+              name: sessionIndexTemplateName,
+            });
+            this.options.logger.debug('Successfully deleted session legacy index template.');
+          } catch (err) {
+            this.options.logger.error(
+              `Failed to delete session legacy index template: ${err.message}`
+            );
+            return reject(err);
+          }
+        }
```

Comment:
Do we really want these errors to fail session index initialization? Since this is a cleanup of an old/unused template, would it be safe to continue with the init phase?

If not, then I think we still have a race condition to contend with: if multiple Kibana instances attempt to delete the legacy template at the same time, the first one will win, and the rest will fail. The instances that fail will be unable to manage sessions until it's restarted, if I understand correctly.

Reply:
> Do we really want these errors to fail session index initialization? Since this is a cleanup of an old/unused template, would it be safe to continue with the init phase?

It's technically safe to proceed, but I opted to consistency since there shouldn't be any "expected" reason why these methods can fail. And if they fail, then most likely the next operations will fail because of the same reason as well (unhealthy or unavailable cluster).

But the main reason why I feel that it's safe to fail is that we [retry to initialize session index until we succeed](https://github.com/elastic/kibana/blob/e0eaca4d768efffb1e81f684cc999dcab191cfa1/x-pack/plugins/security/server/session_management/session_management_service.ts#L89-L95), no restart is needed.

> If not, then I think we still have a race condition to contend with: if multiple Kibana instances attempt to delete the legacy template at the same time, the first one will win, and the rest will fail. The instances that fail will be unable to manage sessions until it's restarted, if I understand correctly.

Yeah, there can be conflicts/409 with the new code as well as there can be conflicts with the existing code (while attempting to create index template) :slightly_smiling_face: Assuming we have no bugs, the second instance will retry initialization after failure and will eventually succeed. Do you see any case where this logic won't apply?

Reply:
> But the main reason why I feel that it's safe to fail is that we retry to initialize session index until we succeed, no restart is needed.

AHH I forgot about this! Sorry, disregard my comments then. I think what you have is great 🙂